data/carepartner/alerts/alerts.go

```diff
@@ -0,0 +1,85 @@
+package alerts
+
+import (
+	"net/http"
+
+	"github.com/tidepool-org/platform/data/service"
+	"github.com/tidepool-org/platform/log"
+)
+
+// List the care partner alerts.
+func List(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	lgr := log.LoggerFromContext(ctx)
+
+	lgr.Info("list the care partner alerts")
+	sCtx.RespondWithStatusAndData(http.StatusNotImplemented, nil)
+}
+
+// Get a specific care partner alert.
+func Get(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	lgr := log.LoggerFromContext(ctx)
+
+	lgr.Info("get a specific care partner alert")
+	sCtx.RespondWithStatusAndData(http.StatusNotImplemented, nil)
+}
+
+// Create a new care partner alert.
+func Create(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	lgr := log.LoggerFromContext(ctx)
+
+	lgr.Info("create a care partner alert")
+	sCtx.RespondWithStatusAndData(http.StatusNotImplemented, nil)
+}
+
+// Update a care partner alert.
+func Update(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	lgr := log.LoggerFromContext(ctx)
+
+	lgr.Info("update a care partner alert")
+	sCtx.RespondWithStatusAndData(http.StatusNotImplemented, nil)
+}
+
+// Ack acknowledges an alert.
+func Ack(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	lgr := log.LoggerFromContext(ctx)
+
+	lgr.Info("create an alert acknowledgement")
+	sCtx.RespondWithStatusAndData(http.StatusNotImplemented, nil)
+}
+
+// CreateAppleTokens stores a token used to push alerts to Apple devices.
+func CreateAppleToken(sCtx service.Context) {
+	ctx := sCtx.Request().Context()
+	lgr := log.LoggerFromContext(ctx)
+
+	lgr.Info("store an Apple push notification token")
+	sCtx.RespondWithStatusAndData(http.StatusNotImplemented, nil)
+}
+
+// AuthWrapper lets the base service inject authentication.
+//
+// Cuz alerts don't care how users are authenticated.
+type AuthWrapper = func(service.HandlerFunc) service.HandlerFunc
+
+// PathBuilder builds a path whose root is pre-determined.
+type PathBuilder = func(segments ...string) string
```

Comment:
I'm not a big fan of this. We have other examples of a similar pattern which makes searching for env variables impossible (segments are underscores, so you can't just grep to find all uses of an env var). 

I think readability is more important, especially in this codebase which has a lot of abstraction and indirection, so I'd stick to static paths.

Reply:
I understand where you're coming from with the grepping... I was thinking it might help avoid a situation where two packages are trying to register the same route. Maybe the routing library takes care of that, or maybe it just says "last to register wins", or who knows. Those kinds of bugs can be a real pain, and with the static routes that are in there now, most of the routes are hidden, ie the `DataSetRoutes`, `SourcesRoutes`, etc are off somewhere else, and point who knows where.

So for getting a view of which endpoints go where, it might almost be better to use the kind of scheme I've proposed because then data/service/api/v1/v1.go would at least be able to tell you where each package is rooted. The grepping is still an issue I guess, if you happen to grep for a complete route, but grepping for `/v1/carepartners/` for example would lead you to the v1.go file, where you could see that every route under that prefix is handled by the `alerts` package.

I'm not hard set on the idea, so I'll follow the existing pattern if after considering the above you aren't convinced. :)

Reply:
After more thought, I'm gonna just do static routes as you've asked. I was on the fence as to whether it was over complicating things or not, and your comments are a clear indicator that yes, it is.

Reply:
The `/v1/carepartner` prefix will never change for those routes, so there's no need build those dynamically. I think this is a code smell. 

You don't need to add your route table in `/data/service/api/v1`. You can define your routes in  `/carepartner/api/v1.go` and include those in `data/service/api/standard.go`. This way "data" related routes and "alerts" related routes will be separate.  

> I was thinking it might help avoid a situation where two packages are trying to register the same route.

When you don't have a top-level view of all routes and those are built dynamically it's easier to define a duplicate elsewhere because you can't even grep for the full route. If duplicate routes are a problem we should make the router panic when a duplicate is added.

Reply:
> The /v1/carepartner prefix will never change for those routes, so there's no need build those dynamically. I think this is a code smell.

It was never about building them dynamically, it was about safely delegating control to divide the routing space. The v1 router could say, "OK carepartner, here's your slice of the routing domain, do with it what you will", while knowing that it couldn't do anything but generate routes under /v1/carepartner, which would allow the v1 package to ensure that there wouldn't be conflicts.

With the current setup, the carepartner package can add routes wherever it likes and there can be collisions. But if all the routes are defined clearly in v1 I don't see it as a problem. That said, I don't think they are at the moment (see DataSetRoutes, SourceRoutes and friends). If/when they are, it could still be a potential problem, but at least it should be easily found, and you're right, it's overall simpler to just have all the routes in one place, and greppable. That's good value. I'm for that.

> When you don't have a top-level view of all routes and those are built dynamically it's easier to define a duplicate elsewhere because you can't even grep for the full route. If duplicate routes are a problem we should make the router panic when a duplicate is added.

I don't see that it exists now (again see DataSetRoutes, SourceRoutes, etc), and from my reading, the router is from go-json-rest, so modifying it to panic it probably not a realistic option. I checked, and it says:

> // This validates the Routes and prepares the Trie data structure.
> // It must be called once the Routes are defined and before trying to find Routes.
> // The order matters, if multiple Routes match, the first defined will be used.

So there will be no warnings when multiple routes match.

IMO, one doesn't need a view of every route in existence in once place, so long as the trail is clearly marked, ie v1 delegates control to sub packages that can't add routes outside of their delegated prefixes.

But again, I agree with you that it's simpler to just have all the routes in v1 (or somewhere) clearly visible and searchable. So I'll redesign to that. :)